### ISSUE TITLE
Possible race condition regarding env_build.xml during build

### ISSUE THREAD

**Issue by mnlevy1981** [ESMCI/cime#2448]
This is everyone's favorite type of issue: an intermittent error that seems to be related to doing things in parallel (and so far has only been seen on cheyenne though there's no indication that the machine itself is causing problems)

**The Issue**

When running `create_test` from a compute node (with `--parallel-jobs 4 --proc-pool 36`), I will occasionally get build errors that trace back to ww3 not being able to build because of issues reading `env_build.xml`. These occur during `C` or `G` compset tests. and @klindsay28 has also run into this problem (although I am unsure of his settings for `parallel-jobs` and `proc-pool`).

@billsacks and @sherimickelson have gotten the same error, but via a different mechanism. Both of them report that trying to build on the cheyenne head-node (which is frowned upon by CISL and typically killed before completion) and then trying to build again from a compute node without running `./case.build --clean-all` in between will then generate this.

**The error**

The `wav.bldlog` will contain something like

```
ERROR: Command: '/usr/bin/xmllint --noout --schema /glade/p/work/mlevy/codes/CESM/cesm2_0_beta09+pop2_trunk/cime/config/xml_schemas/env_entry_id.xsd /glade2/scratch2/mlevy/tests/aux_pop_MARBL/cime5.4.0-alpha.23-cice5_20180123-cesm_pop_2_1_20180406/ERS.T62_g37.G1850ECO.cheyenne_intel.pop-cice_ciso.G.20180410_144957_n54xs4/env_build.xml' failed with error '/glade2/scratch2/mlevy/tests/aux_pop_MARBL/cime5.4.0-alpha.23-cice5_20180123-cesm_pop_2_1_20180406/ERS.T62_g37.G1850ECO.cheyenne_intel.pop-cice_ciso.G.20180410_144957_n54xs4/env_build.xml:1: parser error : Document is empty
```

or

```
ERROR: Command: '/usr/bin/xmllint --noout --schema /glade/p/work/klindsay/cesm2_sandbox/cime/config/xml_schemas/env_entry_id.xsd /glade2/scratch2/klindsay/tests/cesm2_0_beta09_marbl_dev_klindsay_n116.a02_intel_cheyenne_aux_pop_MARBL/ERS.T62_g37.G1850ECO.cheyenne_intel.pop-cice_ecosys_dust_flux_driver.GC.20180405_101112_cx86jm/env_build.xml' failed with error '/glade2/scratch2/klindsay/tests/cesm2_0_beta09_marbl_dev_klindsay_n116.a02_intel_cheyenne_aux_pop_MARBL/ERS.T62_g37.G1850ECO.cheyenne_intel.pop-cice_ecosys_dust_flux_driver.GC.20180405_101112_cx86jm/env_build.xml:195: parser error : expected '>'
      <valid_values>TRUE,FALSE</valid_va
```

**Hypothesis**

Some components (namely POP and CICE) write to `env_build.xml` during the build process. It seems likely that, given the parallelization of the build, ww3 is trying to read `env_build.xml` WHILE another component is writing it.

**Potential Fixes**

Two fixes come to mind:

1. `xmlchange` would create a lock file before changing any xml files, and then remove the lock file when the change has been made. Both `xmlchange` and `xmlquery` would look for a lock file before trying to open the xml file and, if one is found, wait for it to be erased before proceeding
   * PRO: will apply to all uses of `xmlchange` and `xmlquery`, not just those in the build scripts; will apply to all xml files, not just `env_build.xml`
   * CON: reduces the window of potential conflict between two processes but doesn't eliminate it entirely. (There will be some number of flops between `xmlchange` looking for the lock file and creating it, and therefore a very small chance that `xmlquery` looks for the lock file in that intervening time and then opens the file while `xmlchange` is changing it).
1. Instead of calling `xmlchange` directly from component scripts, the components could generate a `[COMP].xmlchange` file containing all the requested changes. In the linking stage, right before changing `BUILD_COMPLETE=TRUE`, the `case.build` could execute all the changes requested by all the components.
   * PRO: No chance of this race condition entering
   * CON: Other instances of `xmlquery` and `xmlchange` running simultaneously could be introduced later in other stages of CIME; if components are writing a value to `env_build.xml` and then relying on it later, those scripts would need to be changed (perhaps by also writing to `config_cache.xml`?)

I don't have a preference between the two proposed fixes, though perhaps implementing both would be safest?

**Comment by jedwards4b** [comment]
@jgfouca has already thought a lot about this kind of race condition and should have input on this issue.

**Comment by billsacks** [comment]
I just ran into this issue with a rebuild failing in ww3. In my case, I had successfully built the case, then reran ./case.build on the login node, when the build shouldn't have done anything meaningful because the model was already successfully built. The build failed in ww3 the first two times around, then failed in the third. Rebuilding like this could be a good way to test a fix (assuming the problem is the same), because – based on this experience – it seems like a fast rebuild often triggered this problem.

First failure:

```
Traceback (most recent call last):
  File "./xmlquery", line 361, in <module>
    _main_func()
  File "./xmlquery", line 284, in _main_func
    with Case(caseroot) as case:
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/case.py", line 83, in __init__
    self.read_xml()
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/case.py", line 192, in read_xml
    self._env_entryid_files.append(EnvBuild(self._caseroot, components=components))
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/env_build.py", line 17, in __init__
    EnvBase.__init__(self, case_root, infile, schema=schema)
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/env_base.py", line 21, in __init__
    EntryID.__init__(self, fullpath, schema=schema, read_only=False)
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/entry_id.py", line 17, in __init__
    GenericXML.__init__(self, infile, schema, read_only=read_only)
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/generic_xml.py", line 54, in __init__
    self.read(infile, schema)
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/generic_xml.py", line 82, in read
    self.read_fd(fd)
  File "/glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/scripts/Tools/../../scripts/lib/CIME/XML/generic_xml.py", line 106, in read_fd
    self.tree = ET.parse(fd)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1182, in parse
    tree.parse(source, parser)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 657, in parse
    self._root = parser.close()
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1654, in close
    self._raiseerror(v)
  File "/usr/lib64/python2.7/xml/etree/ElementTree.py", line 1506, in _raiseerror
    raise err
xml.etree.ElementTree.ParseError: no element found: line 1, column 0
gmake: /Makefile: No such file or directory
gmake: *** No rule to make target '/Makefile'.  Stop.
ERROR: gmake complib -j 8 MODEL=ww3 COMPLIB=/glade/scratch/hannay/my_291_2/bld/lib/libwav.a -f /Makefile MACFILE=/glade/scratch/sacks/cesm_cases/my_291_2/Macros.cheyenne failed: 512
```

Second failure:

```
ERROR: Command: '/usr/bin/xmllint --noout --schema /glade2/scratch2/sacks/cesm_code/cesm2_0_exp10d/cime/config/xml_schemas/env_entry_id.xsd /glade2/scratch2/sacks/cesm_cases/my_291_2/env_build.xml' failed with error '/glade2/scratch2/sacks/cesm_cases/my_291_2/env_build.xml:1: parser error : Document is empty

^' from dir '/glade2/scratch2/sacks/cesm_cases/my_291_2'
gmake: /Makefile: No such file or directory
gmake: *** No rule to make target '/Makefile'.  Stop.

```

**Comment by jedwards4b** [comment]
As far as I can tell the only one writing to env_build.xml is pops build-namelist where it sets the value of CPL_I2O_PER_CAT.   We can greatly reduce the chances of this problem occurring by first checking the current value of this variable and only changing it if it needs to be changed.

**Comment by mnlevy1981** [comment]
POP's `buildcpp` contains

```
$ grep case.set_value buildcpp
                case.set_value("MARBL_NT", MARBL_NT)
            case.set_value("POP_BLCKX", config[2])
            case.set_value("POP_BLCKY", config[3])
            case.set_value("POP_MXBLCKS",config[4])
            case.set_value("POP_DECOMPTYPE", config[5])
            case.set_value("POP_NX_BLOCKS", config[6])
            case.set_value("POP_NY_BLOCKS", config[7])
    pop_cppdefs = case.set_value("POP_CPPDEFS", pop_cppdefs)
```

And also a `case.flush()` -- don't those combine to regenerate the XML files?

**Comment by jedwards4b** [comment]
They do, but in the case outlined by @billsacks above I don't believe that code is being invoked.   The build-namelist issue I pointed out is a problem that should be fixed whether or not it is a complete resolution to this problem.

**Comment by jedwards4b** [comment]
@mnlevy1981 is the issue in pops build-namelist scheduled to be fixed before the release?

**Comment by mnlevy1981** [comment]
Should this go into POP, or does it make more sense for `xmlchange` to determine if it is actually changing anything? It seems like it would be better if `xmlchange` leaves the file untouched unless something is actually being altered? (I'll talk to @alperaltuntas about getting the fix in on the POP side as well)

**Comment by jedwards4b** [comment]
@mnlevy1981 that was an excellent suggestion - implemented in #2567

**Comment by mvertens** [comment]
This is mostly addressed by PR #2567 and so will be closed for now - unless this issue seems to come up again.
